test/widgets/compose_box_test.dart

```diff
@@ -1407,38 +1412,158 @@ void main() {
+        narrow: topicNarrow,
+        isChannelSubscribed: true,
+        channelPostPolicy: ChannelPostPolicy.moderators,
+        selfUserRole: UserRole.administrator,
+        expected: true);
+
+      doTestComposeBoxShown(
+        narrow: topicNarrow,
+        isChannelSubscribed: false,
+        channelPostPolicy: ChannelPostPolicy.moderators,
+        selfUserRole: UserRole.administrator,
+        expected: true);
+
+      doTestComposeBoxShown(
+        narrow: channelNarrow,
+        isChannelSubscribed: true,
+        channelPostPolicy: ChannelPostPolicy.administrators,
+        selfUserRole: UserRole.moderator,
+        expected: false);
+
+      doTestComposeBoxShown(
+        narrow: channelNarrow,
+        isChannelSubscribed: false,
+        channelPostPolicy: ChannelPostPolicy.administrators,
+        selfUserRole: UserRole.moderator,
+        expected: false);
+
+      doTestComposeBoxShown(
+        narrow: topicNarrow,
+        isChannelSubscribed: true,
+        channelPostPolicy: ChannelPostPolicy.administrators,
+        selfUserRole: UserRole.moderator,
+        expected: false);
+
+      doTestComposeBoxShown(
+        narrow: topicNarrow,
+        isChannelSubscribed: false,
+        channelPostPolicy: ChannelPostPolicy.administrators,
+        selfUserRole: UserRole.moderator,
+        expected: false);
+
+      void doTestRefreshSubscribeButtons({required Narrow narrow}) {
+        testWidgets('Refresh/Subscribe buttons when cannot send and channel unsubscribed, $narrow', (tester) async {
+          final channel = eg.stream(streamId: 1,
+            channelPostPolicy: ChannelPostPolicy.administrators);
+          final messages = List.generate(100, (i) => eg.streamMessage(id: 1000 + i,
+            stream: channel, topic: topicNarrow.topic.apiName));
 
-        testWidgets('error banner is shown in $narrowType narrow', (tester) async {
           await prepareComposeBox(tester,
-            narrow: narrow,
-            selfUser: eg.user(role: UserRole.moderator),
-            streams: [eg.stream(streamId: 1,
-              channelPostPolicy: ChannelPostPolicy.administrators)]);
-          checkComposeBox(isShown: false);
+            narrow: ChannelNarrow(channel.streamId),
+            selfUser: eg.user(role: UserRole.member),
+            streams: [channel],
+            subscriptions: [],
+            messages: messages);
+          checkComposeBoxIsShown(false,
+            bannerLabel: zulipLocalizations.composeBoxBannerLabelUnsubscribedWhenCannotSend);
+          check(store.debugMessageListViews).single
+            ..fetched.isTrue()..messages.length.equals(100);
+
+          connection.prepare(json:
+            eg.newestGetMessagesResult(foundOldest: true, messages: messages).toJson(),
+            delay: Duration(seconds: 1));
+          await tester.tap(find.widgetWithText(ZulipWebUiKitButton, 'Refresh'));
+          await tester.pump();
+          check(store.debugMessageListViews).single
```

Comment:
How about:
```suggestion
          final model = MessageListPage.ancestorOf(state.context).model;
          check(model)
```

then reuse `model` below.

(Oh and I guess do this a bit above, at the first reference to the model.)